### PR TITLE
test: integrate 08 wasm etheruem light client into e2e test

### PR DESCRIPTION
**Pull request by gjermundgaraba**
## Description

closes: #143

There is a lot of refactoring done in this, and a lot of deleted code since we moved from using protobuf encoded client and consensus state to json encoded that is defined in the rust code (and generated go types for).

I've gone through all the comments from #150 and sorted out all the issues that I commented would be done in #143. The rest will be done in the issues mentioned in each comment.

The only thing that is missing e2e and fixture testing is timeout/non-membership. I will do that as part of #144, but as a small separate PR. For now, this works with our current e2e test suite.

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Linked to GitHub issue with discussion and accepted design, OR link to spec that describes this work.
- [ ] Wrote unit and integration tests.
- [ ] Added relevant natspec and `godoc` comments.
- [ ] Provide a [conventional commit message](https://www.conventionalcommits.org/en/v1.0.0/) to follow the repository standards.
- [ ] Re-reviewed `Files changed` in the GitHub PR explorer.
- [ ] Review `SonarCloud Report` in the comment section below once CI passes.

### CODE REVIEW

**Review by srdtrk — COMMENTED** [review]
Looking good. I added some code quality suggestions, I think we can merge this soon once those are addressed. I think the PR can be renamed to a feature since it has a lot of non-test changes. Perhaps `feat: working ethereum light client`?

File: packages/tree_hash/src/impls.rs
Comment (gjermundgaraba) [reply packages/tree_hash/src/impls.rs:314]: I fixed that yesterday and forgot to push :(

File: e2e/interchaintestv8/types/ethereum/types.gen.go
Comment (gjermundgaraba) [reply e2e/interchaintestv8/types/ethereum/types.gen.go:10]: It is just to export all the types. I will rename and comment to make that clearer.

File: e2e/interchaintestv8/proto/union/ibc/lightclients/ethereum/v1/ethereum.proto
Comment (gjermundgaraba) [reply e2e/interchaintestv8/proto/union/ibc/lightclients/ethereum/v1/ethereum.proto:1]: I can see that point. I'll add a task to look into this to the epic and we can circle back to that. The main reason I went this route was to avoid having to define the types multiple places, and I don't think (but will be happy to be wrong about this) we can get the alloy types (B256, Bloom, etc) generated directly from that. Which means we would have to have define another struct that translates the proto generated type into a type we use inside the light client (this is what union does, and I don't like it).

Another upside of using protobuf is the consensus state size is probably smaller than the json encoded version.

Comment (gjermundgaraba) [reply e2e/interchaintestv8/proto/union/ibc/lightclients/ethereum/v1/ethereum.proto:1]: #184

File: justfile
Comment (gjermundgaraba) [reply justfile:92]: Added this comment as reference in the issue where we can consider this further: #184

File: packages/ethereum-light-client/src/verify.rs
Comment (gjermundgaraba) [reply packages/ethereum-light-client/src/verify.rs]: Yeah, I removed them as well yesterday and forgot to push. Should be fixed now.

File: packages/ethereum-types/src/consensus/light_client_header.rs
Comment (gjermundgaraba) [reply packages/ethereum-types/src/consensus/light_client_header.rs:32]: I actually did this on purpose to avoid having to add serde_with as a dependency for a single type. I am also OK with adding it, but just seems unnecessary :)

File: packages/utils/Cargo.toml
Comment (gjermundgaraba) [reply packages/utils/Cargo.toml:1]: I am OK with that. I just figured since they were not specific anymore, but if you want to not have those as general utils, I can rename it 👍

File: packages/ethereum-light-client/Cargo.toml
Comment (gjermundgaraba) [reply packages/ethereum-light-client/Cargo.toml:31]: I've seen both ways used with dependencies, and it does work, but using `dep` does make it clearer that it is a dependency on a crate and not another feature. Will change and move it.

File: packages/ethereum-light-client/src/bin/generate_json_schema.rs
Comment (gjermundgaraba) [reply packages/ethereum-light-client/src/bin/generate_json_schema.rs:1]: We'll look into it in #184 👍

File: packages/ethereum-light-client/src/client_state.rs
Comment (gjermundgaraba) [reply packages/ethereum-light-client/src/client_state.rs:18]: It is. I added it as a comment on the consensus state timestamp, will add it here too.

File: packages/ethereum-light-client/src/client_state.rs
Comment (gjermundgaraba) [reply packages/ethereum-light-client/src/client_state.rs:14]: Yeah, the type deserializes it from hex now.

File: packages/ethereum-types/Cargo.toml
Comment (gjermundgaraba) [reply packages/ethereum-types/Cargo.toml:1]: Yeah, will create an issue for that.

File: packages/ethereum-light-client/src/header.rs
Comment (gjermundgaraba) [reply packages/ethereum-light-client/src/header.rs]: Yeah, this was also fixed yesterday and forgot to push. Should be OK now.

File: packages/ethereum-types/src/consensus/light_client_header.rs
Comment (srdtrk) [reply packages/ethereum-types/src/consensus/light_client_header.rs:32]: I think you use this in other places as well

File: packages/ethereum-light-client/src/header.rs
Comment (gjermundgaraba) [reply packages/ethereum-light-client/src/header.rs:49]: I just removed Default from here and the consensus state. It was just used to make the tests a little simpler. I've kept it for ClientState just because it doesn't require anything extra and make some test shorter ^^

File: packages/ethereum-light-client/src/header.rs
Comment (gjermundgaraba) [reply packages/ethereum-light-client/src/header.rs:1]: No, these types are specific to the light client only.

File: packages/ethereum-light-client/src/lib.rs
Comment (gjermundgaraba) [reply packages/ethereum-light-client/src/lib.rs:19]: It is used to have the test-utils available to the cosmswasm package.

File: packages/ethereum-light-client/src/test_utils/bls_verifier.rs
Comment (gjermundgaraba) [reply packages/ethereum-light-client/src/test_utils/bls_verifier.rs:21]: Yeah, would love that, go for it!

File: packages/ethereum-light-client/src/membership.rs
Comment (gjermundgaraba) [reply packages/ethereum-light-client/src/membership.rs:45]: Oh yeah, that seems unnecessary. Fixing it.

File: packages/ethereum-light-client/src/update.rs
Comment (gjermundgaraba) [reply packages/ethereum-light-client/src/update.rs:67]: Not necessarily, hence the comment above. I am not sure when this would ever be used, and I want to look into removing this functionality unless we need it.

File: packages/ethereum-light-client/src/update.rs
Comment (gjermundgaraba) [reply packages/ethereum-light-client/src/update.rs:19]: Agreed, fixed :)

File: packages/ethereum-light-client/src/update.rs
Comment (gjermundgaraba) [reply packages/ethereum-light-client/src/update.rs:68]: Yeah, that is clearer.

File: packages/ethereum-types/Cargo.toml
Comment (gjermundgaraba) [reply packages/ethereum-types/Cargo.toml:1]: #185

File: packages/ethereum-light-client/src/update.rs
Comment (gjermundgaraba) [reply packages/ethereum-light-client/src/update.rs]: Yeah! That is nice, adding it!

File: packages/ethereum-light-client/src/update.rs
Comment (gjermundgaraba) [reply packages/ethereum-light-client/src/update.rs:52]: This doesn't work anymore after consuming the arguments

File: packages/ethereum-light-client/src/update.rs
Comment (gjermundgaraba) [reply packages/ethereum-light-client/src/update.rs]: Only one block down, which doesn't make it much cleaner. With the latest changes, I think maybe you can take another look at this function and see what makes sense now. Probably some things I still missed :)

File: packages/ethereum-types/src/consensus/light_client_header.rs
Comment (gjermundgaraba) [reply packages/ethereum-types/src/consensus/light_client_header.rs:32]: Not serde_with. This particular serde module, yes, but it is the only thing.

File: packages/solidity/Cargo.toml
Comment (gjermundgaraba) [reply packages/solidity/Cargo.toml:10]: Haha, yes, I forgot to remove it from this crate while I was testing. Removed now :)

File: packages/ethereum-types/README.md
Comment (gjermundgaraba) [reply packages/ethereum-types/README.md:1]: Done

File: packages/ethereum-light-client/src/update.rs
Comment (srdtrk) [reply packages/ethereum-light-client/src/update.rs]: I actually ended up cloning again to seperate `new_consensus_state` from `current_consensus_state`

File: packages/ethereum-light-client/src/test_utils/bls_verifier.rs
Comment (srdtrk) [reply packages/ethereum-light-client/src/test_utils/bls_verifier.rs:21]: done

File: packages/ethereum-types/src/consensus/light_client_header.rs
Comment (srdtrk) [reply packages/ethereum-types/src/consensus/light_client_header.rs]: I think introducing this dep is better than maintaining a function in our codebase

**Review by srdtrk — APPROVED** [review]
Lgtm. But I have one more question before merging this. I can't actually see where the proof / signature verification is being made in `update_state` or `update_consensus_state`?

**Comment by gjermundgaraba** [comment]
> Lgtm. But I have one more question before merging this. I can't actually see where the proof / signature verification is being made in `update_state` or `update_consensus_state`?

It doesn't need to, because update_state is being called after verify_client_message, so it is not necessary to do again. If you look at how 07-tendermint in ibc-go handles this, it also does the same.

**Pull request merged**
